dd-trace-core/src/main/java/datadog/trace/core/scopemanager/ContinuableScopeManager.java

```diff
@@ -108,47 +121,66 @@ public void addScopeListener(final ScopeListener listener) {
     scopeListeners.add(listener);
   }
 
-  private class ContinuableScope extends DelegatingScope implements Scope {
+  protected ScopeStack scopeStack() {
+    return this.tlsScopeStack.get();
+  }
+
+  private static final class ContinuableScope extends DelegatingScope implements Scope {
+    private final ContinuableScopeManager scopeManager;
+
     /** Scope to placed in the thread local after close. May be null. */
-    private final ContinuableScope toRestore;
+    // private final ContinuableScope toRestore;
+
     /** Continuation that created this scope. May be null. */
     private final ContinuableScopeManager.Continuation continuation;
     /** Flag to propagate this scope across async boundaries. */
     private final AtomicBoolean isAsyncPropagating = new AtomicBoolean(false);
-    /** depth of scope on thread */
-    private final int depth;
 
     private final ScopeSource source;
 
     private final AtomicInteger referenceCount = new AtomicInteger(1);
 
     ContinuableScope(
+        final ContinuableScopeManager scopeManager,
         final ContinuableScopeManager.Continuation continuation,
         final Scope delegate,
         final ScopeSource source) {
       super(delegate);
       assert delegate.span() != null;
+      this.scopeManager = scopeManager;
       this.continuation = continuation;
-      toRestore = tlsScope.get();
-      depth = toRestore == null ? 0 : toRestore.depth() + 1;
       this.source = source;
     }
 
     @Override
     public void close() {
-      if (referenceCount.decrementAndGet() > 0) {
-        return;
-      }
-
-      if (tlsScope.get() != this) {
-        log.debug("Tried to close {} scope when {} is on top. Ignoring!", this, tlsScope.get());
+      ScopeStack scopeStack = scopeManager.scopeStack();
+
+      // DQH - Aug 2020 - Preserving our broken reference counting semantics for the
+      // first round of out-of-order handling.
+      // When reference counts are being used, we don't check the stack top --
+      // so potentially we undercount errors
+      // Also we don't report closed until the reference count == 0 which seems
+      // incorrect given the OpenTracing semantics
+      // Both these issues should be corrected at a later date
+
+      boolean alive = decrementReferences();
+      if (alive) return;
+
+      boolean onTop = scopeStack.checkTop(this);
+      if (!onTop) {
+        if (log.isDebugEnabled()) {
+          // Using noFixupTop because I don't want to have code with side effects in logging code
+          log.debug(
+              "Tried to close {} scope when not on top. Ignoring!", this, scopeStack.noFixupTop());
```

Comment:
There's only one template in the pattern, but two parameters

```suggestion
              "Tried to close {} scope when not on top. Ignoring!", scopeStack.noFixupTop());
```